src/System.Buffers.Experimental/System/Buffers/MemoryList.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Sequences;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers
+{
+    public class MemoryList : IMemoryList<byte>
+    {
+        private Memory<byte> _data;
+        private MemoryList _next;
+        private long _virtualIndex;
+
+        public MemoryList(Memory<byte> bytes)
+        {
+            _data = bytes;
+            _next = null;
+            _virtualIndex = 0;
+        }
+
+        private MemoryList(Memory<byte> bytes, long runningIndex)
+        {
+            _data = bytes;
+            _virtualIndex = runningIndex;
+        }
+
+        public MemoryList Append(Memory<byte> bytes)
+        {
+            if (_next != null) throw new InvalidOperationException("Node cannot be appended");
+            var node = new MemoryList(bytes, _virtualIndex + _data.Length);
+            _next = node;
+            return node;
+        }
+
+        public Memory<byte> Memory => _data;
+
+        public IMemoryList<byte> Rest => _next;
+
+        public long VirtualIndex => _virtualIndex;
+
+        public int CopyTo(Span<byte> buffer)
+        {
+            int copied = 0;
+            Position position = default;
+            var free = buffer;
+            while (TryGet(ref position, out ReadOnlyMemory<byte> segment, true))
+            {
+                if (segment.Length > free.Length)
+                {
+                    segment.Span.Slice(0, free.Length).CopyTo(free);
```

Comment:
If we had a CopyTo method that returned how much it copied (rather than throwing if not enough space), this could be written more cleanly.

```C#
int copied = 0;
Position position = default;
var free = buffer;
while (TryGet(ref position, out ReadOnlyMemory<byte> segment, true))
{
  copied += segment.Span.CopyTo(free);
  free = buffer.Slice(copied);
  if (free.Length == 0) break;
}
```

@KrzysztofCwalina, does this code motivate the change to our span CopyTo API?

Reply:
Yes